modmail/config.py

```diff
@@ -1,144 +1,740 @@
+    discord_converter: discord.ext.commands.converter.Converter = attr.ib(default=None)
+    discord_converter_attribute: typing.Optional[
+        types.FunctionType
+    ] = None  # if we want an attribute off of the converted value
+
+    # hidden values do not show up in the bot configuration menu
+    hidden: bool = False
+
+    @description.validator
+    def _validate_description(self, attrib: attr.Attribute, value: typing.Any) -> None:
+        if not isinstance(value, attrib.type):  # pragma: no branch
+            raise ValueError(f"description must be of {attrib.type}") from None
+
+    @discord_converter.validator
+    def _validate_discord_converter(self, attrib: attr.Attribute, value: typing.Any) -> None:
+        if value is None:
+            return
+        if not hasattr(value, "convert"):  # pragma: no branch
+            raise AttributeError("Converters must have a method named convert.")
+
+
+@attr.s(auto_attribs=True, slots=True)
+class BotConfig:
+    """
+    Values that are configuration for the bot itself.
+
+    These are metavalues, and are the token, prefix, database bind, basically all of the stuff that needs to
+    be known BEFORE attempting to log in to the database or discord.
+    """
+
+    token: str = attr.ib(
+        default=marshmallow.missing,
+        on_setattr=attr.setters.frozen,
+        repr=False,
+        metadata={
+            "required": True,
+            "dump_only": True,
+            "allow_none": False,
+            METADATA_TABLE: ConfigMetadata(
+                canonical_name="Bot Token",
+                description="Discord bot token. Required to log in to discord.",
+                export_environment_prefill="Bot Token",
+                extended_description="This is obtainable from https://discord.com/developers/applications",
+                export_to_env_template=True,
+            ),
+        },
+    )
+    prefix: str = attr.ib(
+        default="?",
+        metadata={
+            METADATA_TABLE: ConfigMetadata(
+                canonical_name="Command Prefix",
+                description="Command prefix.",
+                export_environment_prefill="?",
+                app_json_default="?",
+                app_json_required=False,
+                export_to_env_template=True,
+            )
+        },
+    )
+    prefix_when_mentioned: bool = attr.ib(
+        default=True,
+        metadata={
+            METADATA_TABLE: ConfigMetadata(description="Use the bot mention as a prefix."),
+        },
+    )
+
+
+@attr.s(auto_attribs=True, slots=True, frozen=True)
+class BotModeConfig:
+    """
+    The three bot modes for the bot. Enabling some of these may enable other bot features.
+
+    `production` is used internally and is always True.
+    `develop` enables additional features which are useful for bot developers.
+    `plugin_dev` enables additional commands which are useful when working with plugins.
+    """
+
+    production: bool = desert.ib(
```

Comment:
There's not really a difference between the two. The source for desert.ib [is this](https://desert.readthedocs.io/en/stable/_modules/desert.html#ib):

Additionally, this isn't worth a comment, since someone reading the docs of desert shall be able to figure this out.
```py
def field(marshmallow_field: marshmallow.fields.Field, **kw) -> dataclasses.Field:
    """Specify a marshmallow field in the metadata for a ``dataclasses.dataclass``.

    .. code-block:: python

        @dataclasses.dataclass
        class A:
            x: int = desert.field(marshmallow.fields.Int())
    """
    meta = metadata(marshmallow_field)
    meta.update(kw.pop("metadata", {}))
    return dataclasses.field(**kw, metadata=meta)
```